### PR TITLE
Add RFS GPIO support

### PR DESCRIPTION
**Pull request by kjbracey-arm**
Add support for triggering RFS via a GPIO line.

### CODE REVIEW

File: edge-client/reset_factory_settings.c [hunk 1]
```diff
@@ -132,6 +139,90 @@ EDGE_LOCAL void rfs_reset_factory_settings_request_cb(void *arg)
...
+#define RFS_GPIO_POLARITY 0
+#endif
+
+static void *rfs_gpio_thread(void *arg)
+{
+    struct gpiod_line *line = arg;
+    struct gpiod_line_event last_event = { 0 };
+    last_event.event_type = GPIOD_LINE_EVENT_FALLING_EDGE ^ RFS_GPIO_POLARITY;
+
+    for (;;) {
```
Comment (JanneKiiskila) [inline edge-client/reset_factory_settings.c:156]: Does this block everything else while it's looping?

File: edge-client/reset_factory_settings.c [hunk 2]
```diff
@@ -132,6 +139,90 @@ EDGE_LOCAL void rfs_reset_factory_settings_request_cb(void *arg)
...
+        tr_info("rfs button %s", event.event_type == GPIOD_LINE_EVENT_RISING_EDGE ? "pressed", "released");
+        // Look for release after press
+        if (last_event.event_type == GPIOD_LINE_EVENT_RISING_EDGE && event.event_type == GPIOD_LINE_EVENT_FALLING_EDGE) {
+            // Check if it was pressed for at least 2 seconds
+            long long secs = event.ts.tv_sec - last_event.ts.tv_sec;
+            if (event.ts.tv_nsec < last_event.ts.tv_nsec) {
+                secs--;
+            }
+            // If so, request the reset, and exit the GPIO handler
+            if (secs >= 2) {
```
Comment (JanneKiiskila) [inline edge-client/reset_factory_settings.c:177]: Should the # of seconds be also configurable? Should it be 10?

File: edge-client/reset_factory_settings.c (hunk 1 above)
Comment (kjbracey-arm) [reply edge-client/reset_factory_settings.c:154]: It's a whole dedicated thread for monitoring the GPIO line.

So this thread will spend all its life blocking in `gpiod_line_event_read`.

File: edge-client/reset_factory_settings.c (hunk 2 above)
Comment (kjbracey-arm) [reply edge-client/reset_factory_settings.c:174]: Certainly could make it configurable. No idea what the default will be.

Nothing will happen until it's released, so I had a vague feeling that having an excessively-long press requirement might be counter-intuitive. Hold for 5 seconds, and nothing happens, would you try again for 10?

Comment (kivaisan) [reply edge-client/reset_factory_settings.c:174]: 2 secs sounds too short for accidental resets. Typically time is ~30sec in consumer devices, but I don't know what kind of requirement this kind of devices has.

**Comment by JanneKiiskila** [comment]
I think 10 seconds is quite common. 30 s is too long, it's a pain to keep it pressed that long.

File: edge-client/reset_factory_settings.c (hunk 2 above)
Comment (kjbracey-arm) [reply edge-client/reset_factory_settings.c]: Changed default to 10, and made it configurable.

Comment (JanneKiiskila) [reply edge-client/reset_factory_settings.c]: Excellent!

File: edge-client/reset_factory_settings.c (hunk 1 above)
Comment (JanneKiiskila) [reply edge-client/reset_factory_settings.c:158]: Good!

File: edge-client/reset_factory_settings.c [hunk 3]
```diff
@@ -132,6 +139,91 @@ EDGE_LOCAL void rfs_reset_factory_settings_request_cb(void *arg)
...
+        // Wait for and read an event
+        if (gpiod_line_event_read(line, &event)) {
+            tr_err("gpiod_line_event_read error");
+            break;
+        }
+        // If unknown event, ignore
+        if (event.event_type != GPIOD_LINE_EVENT_FALLING_EDGE && event.event_type != GPIOD_LINE_EVENT_RISING_EDGE) {
+            continue;
+        }
+        tr_info("rfs button %s", event.event_type == GPIOD_LINE_EVENT_RISING_EDGE ? "pressed", "released");
```
Comment (kivaisan) [inline edge-client/reset_factory_settings.c:169]:
```suggestion
        tr_info("rfs button %s", event.event_type == GPIOD_LINE_EVENT_RISING_EDGE ? "pressed" : "released");
```

**Review by kivaisan — APPROVED** [review]

**Comment by kjbracey-arm** [comment]
Updated just to take out the CHANGELOG.md edit.

**Review by JanneKiiskila — APPROVED** [review]

**Review by JanneKiiskila — APPROVED** [review]

**Pull request merged**